File: 1-two-sum/two-sum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        // for (int i = 0; i < nums.length - 1; i++) {
        //     for (int j = i + 1; j < nums.length; j++) {
        //         if (nums[i] + nums[j] == target) {
        //             return new int[] {i, j};
        //         }
        //     }
        // }

        // return new int[] {};

        Map<Integer, Integer> numMap = new HashMap<>();

        // for (int i = 0; i < nums.length; i++) {
        //     numMap.put(nums[i], i);
        // }

        // for (int i = 0; i < nums.length; i++) {
        //     int complement = target - nums[i];

        //     if (numMap.containsKey(complement) && numMap.get(complement) != i) {
        //         return new int[] {i, numMap.get(complement)};
        //     }
        // }

        // return new int[] {};

        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];

            if (numMap.containsKey(complement)) {
                return new int[] {i, numMap.get(complement)};
            }

            numMap.put(nums[i], i);
        }

        return new int[]{};


        
    }
}

File: 1-two-sum/two-sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # nums = 2, 7, 11, 15
        # target = 9
        # output = 0, 1

        # nums = 3, 2, 4
        # target = 6
        # output = 1, 2

        # for i in range(len(nums)):
        #     complement = target - nums[i]
        #     for j in range(len(nums)):
        #         if j == i:
        #             continue
        #         if nums[j] == complement:
        #             return [i, j]

        hashmap = {}

        # 2: 0
        # 7: 1
        # 11: 2
        # 15: 3

        for i in range(len(nums)):
            hashmap.update({nums[i]: i})
        
        for i in range(len(nums)):
            complement = target - nums[i]
            
            if complement in hashmap and hashmap[complement] != i:
                return [i, hashmap[complement]]
        
        return []